torch/distributed/_sharding_spec/api.py

```diff
@@ -96,3 +96,173 @@ def placement(self) -> ShardPlacement:
         Retrieves the shard placement.
         """
         return self._placement
+
+class GenericShardingSpec(PlacementSpec):
```

Comment:
Will there be another MeshShardingSpec build on top of this, or do we expect users to configure GenericShardingSpec in a mesh manner if they need mesh?

Reply:
My thought was to only have GenericShardingSpec since it covers the Mesh use case in a much more generic way. I feel it is also simpler to have just two specs, one which is simple (ChunkShardingSpec) and probably covers most use cases and then another one which is more generic and covers everything.

Reply:
Thinking about this more I realized the following. MeshShardingSpec can be applied to any tensor (so can ChunkShardingSpec) but GenericShardingSpec can be only applied to specific tensors since it involves shard offsets and lengths. So you need a different spec for each tensor (in case of GenericShardingSpec) instead of having one spec (in case of MeshShardingSpec) for a bunch of tensors.

Although, I do feel we can probably start out with these two to begin with (Generic and Chunk) and then evaluate if we do indeed need another spec like Mesh.

Reply:
I think `ChunkShardingSpec` and `MeshShardingSpec` have a few advantages over `GenericShardingSpec`, where the former defines a sharding function that can be applied to dynamic shapes, where the latter basically hardcodes a sharding function by enumerating the sharding outputs. 

1) If the shape of an input tensor cannot be predefined (dynamic tensor can be common in NLP), then  `MeshShardingSpec` cannot be replaced by `GenericShardingSpec`.
2) Given a list of devices, actually both `ChunkShardingSpec` and `MeshShardingSpec` can be describes as a generator that yields a stream of shards. This may help save some memory if we have a large number of shards.

Reply:
@SciPioneer Agree with what you mentioned above, although we have a concrete use case for uneven/arbitrary shard sizes which MeshShardingSpec doesn't satisfy and we need something like GenericShardingSpec for this.